### PR TITLE
Feature/chunk to database

### PR DESCRIPTION
**Pull request by lauratheq**
Fixes:
* https://app.clickup.com/t/86c15twu0
* https://app.clickup.com/t/86c15twwk
* https://app.clickup.com/t/86c15tx20
* https://app.clickup.com/t/86c15tx5v
* https://app.clickup.com/t/86c15txbv
* https://app.clickup.com/t/86c15txfp
* https://app.clickup.com/t/86c15txrh
* https://app.clickup.com/t/86c15vggb

### CODE REVIEW

File: src/Entities/Chunk.php [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+    /**
+     * @var string|null
+     */
+    private ?string $group = null;
+
+    /**
+     * @var string|null
+     */
+    private ?string $status = null;
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:43]: Add ENUM to specific the status and always use thtat enum. Increas php version as needed

File: src/Entities/Chunk.php [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+     * Constructor
+     *
+     * @param int|null $chunk_id The chunk ID.
+     */
+    public function __construct( ?int $chunk_id = null ) {
+        if ( empty( $chunk_id ) ) {
+            $this->chunk_id = $this->create_chunk();
+        } else {
+            $this->chunk_id = $chunk_id;
+        }
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:80]: Why is this needed? The logic should more be like "if id is set get chunk from id". To save in the db rename "create" to "save". It should be called explicitly. This is how WC does it for products

File: src/Entities/Chunk.php [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+            $this->fetch_data();
+        }
+
+        // If duration is not yet calculated, calculate it
+        if ( empty( $this->duration ) && ! empty( $this->end ) && ! empty( $this->start ) ) {
+            return $this->calculate_duration();
+        }
+
+        return (float) $this->duration;
+    }
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:242]: Refactor to use datetimeimmutable

File: src/Entities/Chunk.php [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+     * Get the end time
+     *
+     * @return string
+     */
+    public function get_end(): string {
+        if ( ! $this->is_data_fetched ) {
+            $this->fetch_data();
+        }
+        return $this->end;
+    }
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:224]: Datetime

File: src/Entities/Chunk.php [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+     */
+    public function get_formatted_duration(): string {
+        $duration = $this->get_duration();
+
+        if ( 0.0 === $duration ) {
+            return '0s';
+        }
+
+        // Format with microseconds precision
+        return sprintf( '%.6fs', $duration );
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:257]: This does not belong to the class. Since php hast a number_format function there is no need for abstraction. Please remove

File: src/Entities/Chunk.php [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        $end_time   = (float) $this->end;
+        $start_time = (float) $this->start;
+
+        $duration = $end_time - $start_time;
+
+        // Store the calculated duration
+        $this->duration = (string) $duration;
+
+        return $duration;
+    }
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:279]: Refactor to use datetimeimmutable

File: src/Chunker.php [hunk 7]
```diff
@@ -53,92 +39,119 @@ protected function schedule_chunk(array|Iterator $chunkData): void
...
-    abstract function process_chunk_data(Generator $chunkData): void;
+    public function cleanup_chunk_data(): void
+    {
+        /**
+         * Filters the number of days to keep chunk data.
+         *
+         * @param string $interval The interval string (e.g., '-14 days').
+         */
+        $days_interval = apply_filters( 'asp/chunks/cleanup/days', '-14 days' );
+        $cleanup_timestamp = (float) strtotime( $days_interval );
```
Comment (JUVOJustin) [inline src/Chunker.php:132]: Not the best implementation since it can introduce bugs by passing positive values. Only allow filtering the integer. As discussed seconds is best. WordPress offers Constants so devs do not need to calculate

File: src/Chunker.php [hunk 8]
```diff
@@ -53,92 +39,119 @@ protected function schedule_chunk(array|Iterator $chunkData): void
...
+                "DELETE FROM {$this->get_chunks_table_name()} WHERE start < %f",
+                $cleanup_timestamp
+            );
+        } else {
+            $query = $this->db()->prepare(
+                "DELETE FROM {$this->get_chunks_table_name()} WHERE status = %s AND start < %f",
+                $status,
+                $cleanup_timestamp
+            );
+        }
```
Comment (JUVOJustin) [inline src/Chunker.php:153]: Reuse enum pls

File: src/Entities/Chunk.php [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+            'action_id' => '%d'
+        );
+        $allowed_fields = apply_filters( 'asp/chunks/allowed_fields', $allowed_fields );
+        $allowed_fields = array_keys( $allowed_fields );
+        foreach ( $allowed_fields as $field ) {
+            if ( isset( $arguments[ $field ] ) ) {
+                $this->{$field} = $arguments[ $field ];
+                $update_args[ $field ] = $arguments[ $field ];
+            }
+        }
```
Comment (JUVOJustin) [inline src/Entities/Chunk.php:321]: What is that about? Why can i add new fields in the first place?

File: src/Sync.php [hunk 10]
```diff
@@ -124,16 +127,19 @@ public function maybe_trigger_last_in_group(int $action_id): void
...
         }
 
-        // Mark action as complete
-        $this->get_stats()->end_action($action_id);
+        // set the end time of the chunk
+        $action_arguments = $action->get_args();
+        if ( !empty( $action_arguments['chunk_id'] ) ) {
+            $chunk = new Chunk( $action_arguments['chunk_id'] );
+            $chunk->update( [
+                'status' => 'finished',
```
Comment (JUVOJustin) [inline src/Sync.php:135]: ENUM

File: src/Sync.php [hunk 11]
```diff
@@ -153,19 +159,25 @@ public function maybe_trigger_last_in_group(int $action_id): void
...
-
-        // Track action start if it is not the complete action
-        if (!str_contains($action->get_hook(), "/complete")) {
-            $this->get_stats()->add_action($action_id);
+        // set the start time of the chunk
+        $action_arguments = $action->get_args();
+        if ( !empty( $action_arguments['chunk_id'] ) ) {
+            $chunk = new Chunk( $action_arguments['chunk_id'] );
+            $chunk->update( [
+                'status'    => 'started',
```
Comment (JUVOJustin) [inline src/Sync.php:174]: ENUM

File: src/Sync.php [hunk 12]
```diff
@@ -204,28 +216,16 @@ public function handle_exception(int $action_id, Exception $e): void
...
         }
 
-        // Update stats
-        $this->get_stats()->mark_action_as_failed($action_id, $e->getMessage());
+        // set the end time of the chunk
+        $action_arguments = $action->get_args();
+        if ( !empty( $action_arguments['chunk_id'] ) ) {
+            $chunk = new Chunk( $action_arguments['chunk_id'] );
+            $chunk->update( [
+                'status' => 'failed',
```
Comment (JUVOJustin) [inline src/Sync.php:224]: ENUM

File: src/Stats.php [hunk 13]
```diff
@@ -6,137 +6,46 @@
...
+        $query = $this->db()->prepare(
+            "SELECT MIN(start) as sync_start, MAX(end) as sync_end 
+            FROM {$this->get_chunks_table_name()} 
+            WHERE `group` = %s",
+            $this->group_name
+        );
 
-        $this->actions[$id]['status'] = 'failed';
-        $this->actions[$id]['error_message'] = $error_message;
-        $this->save();
```
Comment (JUVOJustin) [inline src/Stats.php:118]: you have added a function inside the chunk class why not using it?

File: src/DB.php [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    private function maybe_create_table( \wpdb $wpdb ): void
+    {
+        $table_name = $wpdb->prefix . $this->table_name_chunks;
+
+        $charset_collate = $wpdb->get_charset_collate();
+        $sql = "CREATE TABLE {$table_name} (
+            id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+            name text NOT NULL,
+            `group` text NOT NULL,
```
Comment (JUVOJustin) [inline src/DB.php:68]: all or none wiht `

**Review by JUVOJustin — CHANGES_REQUESTED** [review]

File: src/Stats.php (hunk 13 above)
Comment (lauratheq) [reply src/Stats.php:118]: Because this is the sync duration for the whole group. With MIN(start) and MAX(end) I get the values I need to calculate the whole sync process. The duration method in the Chunk Entity just calculate the duration for a single chunk.

File: src/Entities/Chunk.php (hunk 9 above)
Comment (lauratheq) [reply src/Entities/Chunk.php:321]: I just wanted to leave the possibility open for other plugins to append stuff to the database table. Maybe there is data missing in an edge case where I want to append it to the chunks. If this possibility is not anticipated, I can simplify the save methods.

File: src/Chunker.php [hunk 15]
```diff
@@ -53,92 +40,118 @@ protected function schedule_chunk(array|Iterator $chunkData): void
...
+    public function cleanup_chunk_data(): void
+    {
+        /**
+         * Filters the number of days to keep chunk data.
+         *
+         * @param int $interval The interval (e.g., 14).
+         */
+        $days_interval = apply_filters( 'asp/chunks/cleanup/days', 14 );
+        $days_interval = (int) $days_interval * DAY_IN_SECONDS;
+        $cleanup_timestamp = (int) time() - $days_interval;
```
Comment (JUVOJustin) [inline src/Chunker.php:132]: Just use 
```php
$interval = apply_filters( 'asp/chunks/cleanup', 14*DAY_IN_SECONDS);
$cleanup_timestamp = (int) time() - $interval;
```

File: src/Chunker.php [hunk 16]
```diff
@@ -53,92 +40,118 @@ protected function schedule_chunk(array|Iterator $chunkData): void
...
+                "DELETE FROM {$this->get_chunks_table_name()} WHERE start < %f",
+                $cleanup_timestamp
+            );
+        } else {
+            $query = $this->db()->prepare(
+                "DELETE FROM {$this->get_chunks_table_name()} WHERE status = %s AND start < %f",
+                $status,
+                $cleanup_timestamp
+            );
+        }
```
Comment (JUVOJustin) [inline src/Chunker.php:153]: use enum. Either pass enum to filter or pass string to filter and parse back to enum. 
```php
ProcessStatus::from($status);
```

File: src/DB.php [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+            `status` ENUM({$enum_definition}) NOT NULL,
+            `data` longtext NOT NULL,
+            `start` decimal(14,4) DEFAULT NULL,
+            `end` decimal(14,4) DEFAULT NULL,
+            PRIMARY KEY (`id`),
+            KEY `status` (`status`),
+            KEY `start` (`start`),
+            KEY `end` (`end`),
+            KEY `action_id` (`action_id`)
+        ) {$charset_collate}";
```
Comment (JUVOJustin) [inline src/DB.php:88]: i like what you tried and appreciate the effort, but this will likely cause errors once the enum options change. I think in this particular case storing the string and also parsing it back is the better option.

Ideally in a perfect world we would have a database migration that we could use to add new enum options as well

Comment (JUVOJustin) [reply src/DB.php:88]: Also i just found that SQlite does not support enums

**Review by JUVOJustin — CHANGES_REQUESTED** [review]

**Review by JUVOJustin — APPROVED** [review]
